src/tweetvalidator/models/clustered_cos_sim_model.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Provides CousterCosSimModel which provides a variety of caracterization options
+for a user's tweet corpus including single-mean and multi-cluster.
+
+Additionally provides similarity_score and infer which take tweets and compare
+with user characterization.
+
+June, 2019
+@author: Joshua Rubin
+"""
+
+import numpy as np
+from .base_model import Model
+
+# This is an implementation of k-means on the hypersphere which uses
+# cosine similarity rather than cartesian distance to cluster.
+# https://github.com/jasonlaska/spherecluster
+from spherecluster import SphericalKMeans
+
+class ClusteredCosSimModel(Model):
+    """ Derives a model that creates a mean or clustered tweet characterizaton.
+        Inference is performed by cosine similarity with a threshold to produce
+        true for inconsistent/fraudulent and false for consistent/authentic.
+        Similarity scores used for this discrimination can optionally take
+        cluster-size into account.  Clustering is performed with spherical
+        k-means (which uses similarity rather than cartesian distance) to
+        properly cluster normalized vectors on the unit-hypersphere.
+    """
+    def __init__(self, embedded_corpus=True, max_clusters=1, verbose=False):
+        """ Requires a pre-embedded corpus passed in through <embedded_corpus>.
+        """
+ 
+        if not embedded_corpus:
+            raise Exception('Sorry, currently requires an embedded_corpus.')
+       
+        self.max_clusters = max_clusters
+        self.embedded_corpus = embedded_corpus
+        self.verbose = verbose
+        # Default setting; change with set_hyperparameter on base-class
+        self.params = { 'threshold': 0.3 }   
+        
+        self.characterization_complete = False
+     
+    def characterize(self, corpus, context_corpus):
+        
+        # Because dataframe holds nested arrays awkwardly
+        corpus = np.asarray([x for x in corpus])
+                   
+        kmeans = SphericalKMeans(n_clusters=self.max_clusters,
+                                 random_state=0).fit(corpus)
+               
+        self.cluster_means = kmeans.cluster_centers_
+    
+        sim_sum            = np.zeros(self.max_clusters)
+        self.cluster_count = np.zeros(self.max_clusters, dtype=np.int)
+        self.cluster_edges = np.ones(self.max_clusters)
+    
+        # Compute characteristics of each cluster
```

Comment:
I think I follow this. You just calc the sim score for all the clusters and the clusters are found for k-means or something?